File: 2-minusy/solution.cpp
#include <iostream>

using namespace std;

bool is_parenthesis_opened = false;

void open_parenthesis(string& result) {
    result += '(';
    is_parenthesis_opened = true;
}

void close_parenthesis(string& result) {
    result += ')';
    is_parenthesis_opened = false;
}

int main() {
    ios_base::sync_with_stdio(0); 
    cin.tie(0);
    cout.tie(0);

    int n;
    cin >> n;
    string result = "";

    char curr_char;
    for (int i = 0; i < n - 1; i++) {
        cin >> curr_char;

        if (curr_char == '+' && !is_parenthesis_opened)
            open_parenthesis(result);

        else if (curr_char == '-' && is_parenthesis_opened)
            close_parenthesis(result);
        
        result += '-';
    }

    if (is_parenthesis_opened) 
        close_parenthesis(result);

    cout << result;
    return 0;
}

File: 2-minusy/solution.py
def close_parenthesis(result):
    global is_parenthesis_opened

    result.append(')')
    is_parenthesis_opened = False
    return result

def open_parenthesis(result):
    global is_parenthesis_opened
    
    result.append('(')
    is_parenthesis_opened = True
    return result


length = int(input())
result, is_parenthesis_opened = [], False


for _ in range(length-1):
    char = input()

    if char == '+' and not is_parenthesis_opened:
        open_parenthesis(result)

    elif char == '-' and is_parenthesis_opened:
        close_parenthesis(result)

    result.append('-')

if is_parenthesis_opened:
    close_parenthesis(result)

print(''.join(result))
